File: 프로그래머스/1/77884. 약수의 개수와 덧셈/약수의 개수와 덧셈.cpp
#include <string>
#include <vector>

using namespace std;

bool isEven(int num) {
    int cnt = 0;
    
    // i * i <= num 을 사용하는 이유:
    //   - i*i == num (즉 i = √num)인 경우에도 i는 num의 약수이므로 반드시 검사해야 함
    //   - i*i < num 으로 작성하면 √num을 놓쳐 제곱수의 약수를 제대로 계산할 수 없음
    //   - 소수 판별 알고리즘 참고 (12977. 소수 만들기)
    for(int i = 1; i * i <= num; i++) {
        if(num % i == 0) {
            int j = num / i;
            
            cnt += 2;
            
            if(i == j) {
                cnt--;
            }
        }
    }
    
    return cnt % 2 == 0;
}

int solution(int left, int right) {
    int answer = 0;
    
    for(int i = left; i <= right; i++) {
        if(isEven(i)) {
            answer += i;
        } else {
            answer -= i;
        }
    }
    
    return answer;
}

File: 프로그래머스/1/77884. 약수의 개수와 덧셈/약수의 개수와 덧셈.java
class Solution {
    public int solution(int left, int right) {
        int answer = 0;
        
        for(int i = left; i <= right; i++) {
            // 약수의 개수가 홀수 = 완전 제곱수 (예: 25, 36)
            int sqrt = (int) Math.sqrt(i);
            
            if(sqrt * sqrt == i) {
                answer -= i;
            } else {
                answer += i;
            }
        }
        
        return answer;
    }
}

/**
 * [ i / sqrt == sqrt 가 적절하지 않은 이유 ]
 *
 * i / sqrt 는 정수 나눗셈으로 수행되기 때문에
 * 나눗셈 결과의 소수점 이하가 모두 버려진다.
 *
 * 이로 인해 완전제곱수 여부를 정확하게 판단할 수 없다.
 *
 * 예시)
 *  - i = 15
 *  - sqrt = (int) Math.sqrt(15) = 3
 *
 *  정수 나눗셈 결과:
 *  - i / sqrt = 15 / 3 = 5
 *
 * 따라서 완전제곱수 판별은
 *  - sqrt * sqrt == i
 * 와 같이 곱셈을 이용한 정수 비교 방식이 안전하고 권장된다.
 */
